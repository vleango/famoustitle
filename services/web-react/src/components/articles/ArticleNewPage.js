import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import { Button } from 'reactstrap';
import { Form, FormGroup, Input } from 'reactstrap';
import { split, map, trim, uniq } from 'lodash';
import { toastInProgress, toastSuccess, toastFail } from '../shared/Toast';

import { createItem } from '../../actions/articles';
import Header from '../shared/headers/Header';

export class ArticleNewPage extends Component {

    constructor(props) {
        super(props);
        this.state = {
            title: "",
            body: "",
            tags: "",
            submitting: false,
            errorMsg: "",
            token: props.token
        };
    }

    onSubmitArticle = async (e) => {
        e && e.preventDefault();
        const { title, body } = this.state;
        if(title === "" || body === "") {
            this.setState({ errorMsg: "title or body is blank" });
            return;
        }

        const toastID = toastInProgress("Saving in progress...");

        try {
            this.setState({ submitting: true, errorMsg: "" });
            const rawTags = split(this.state.tags, ',');
            const trimmedTags = map(rawTags, (tag) => { return trim(tag).toLowerCase() });
            const tags = uniq(trimmedTags);
            const { title, body } = this.state;
            const article = { title, body, tags };
            await this.props.createItem({ token: this.state.token, article });
            toastSuccess("Save successful!", toastID);
            this.props.history.push('/');
        }
        catch(error) {
            this.setState({ submitting: false, errorMsg: "server error" });
            toastFail(e, toastID);
        }
    };

    onInputChange = (e) => {
        const field = e.target.name;
        const value = e.target.value;
        this.setState(() => ({ [field]: value }));
    };

    render() {
        return (
            <Fragment>
                <Header resourceTitle="Create New Article" />
                <div className="container">
                    <Form onSubmit={this.onSubmitArticle} autoComplete="off">
                        <FormGroup>
                            <Input type="text"
                                   name="title"
                                   value={this.state.title}
                                   placeholder="Title"
                                   onChange={this.onInputChange} />
                        </FormGroup>
                        <FormGroup>
                            <Input type="text"
                                   name="tags"
                                   value={this.state.tags}
                                   placeholder="Tag"
                                   onChange={this.onInputChange} />
                        </FormGroup>
                        <FormGroup>
                            <Input type="textarea" rows="20"
                                   name="body"
                                   value={this.state.body}
                                   placeholder="Add your article"
                                   onChange={this.onInputChange} />
                        </FormGroup>

                        { this.state.errorMsg && <p>{this.state.errorMsg}</p> }

                        <div className="clearfix">
                            <Button color="primary float-right" disabled={this.state.submitting} size="lg">Save</Button>
                        </div>
                    </Form>
                </div>
            </Fragment>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        token: state.auth.token
    };
};

const mapDispatchToProps = (dispatch) => ({
    createItem: async (data) => await dispatch(createItem(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(ArticleNewPage);
