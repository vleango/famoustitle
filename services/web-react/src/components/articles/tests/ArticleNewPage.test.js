import React from 'react';
import { shallow } from 'enzyme';
import { ArticleNewPage } from '../ArticleNewPage';

import { createMemoryHistory } from 'history'

describe('Components', () => {
  describe('ArticleNewPage', () => {

    let wrapper = shallow(<ArticleNewPage />);

    describe('Snapshot', () => {
      it('should correctly render ArticleNewPage', () => {
        expect(wrapper).toMatchSnapshot();
      });
    })

    describe('Submit', () => {
      describe('Validation', () => {
        const errorMsg = "title or body is blank";
        describe('Title is blank', () => {
          it('should not submit', () => {
            const target = { title: '', body: 'hello' };
            wrapper.setState({ ...target });
            wrapper.instance().onSubmitArticle(null);
            expect(wrapper.state('errorMsg')).toBe(errorMsg);
            expect(wrapper.state('submitting')).toBe(false);
          });
        });
        describe('Body is blank', () => {
          it('should not submit', () => {
            const target = { title: 'hello', body: '' };
            wrapper.setState({ ...target });
            wrapper.instance().onSubmitArticle(null);
            expect(wrapper.state('errorMsg')).toBe(errorMsg);
            expect(wrapper.state('submitting')).toBe(false);
          });
        });
        describe('Title and Body are blank', () => {
          it('should not submit', () => {
            const target = { title: '', body: '' };
            wrapper.setState({ ...target });
            wrapper.instance().onSubmitArticle(null);
            expect(wrapper.state('errorMsg')).toBe(errorMsg);
            expect(wrapper.state('submitting')).toBe(false);
          });
        });
        describe('Title and Body are present', () => {
          it('should submit', () => {
            const action = jest.fn();
            let wrapper = shallow(<ArticleNewPage createItem={action} history={[]} />);
            const target = { title: 'hello', body: 'bye' };
            wrapper.setState({ ...target });
            wrapper.instance().onSubmitArticle(null);
            expect(action).toHaveBeenCalled();
          });
        });
      });
    });

  });
});
