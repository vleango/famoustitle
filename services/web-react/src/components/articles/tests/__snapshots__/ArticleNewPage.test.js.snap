// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components ArticleNewPage Snapshot should correctly render ArticleNewPage 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ArticleNewPage />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Header
          resourceTitle="Create New Article"
/>,
        <div
          className="container"
>
          <Form
                    autoComplete="off"
                    onSubmit={[Function]}
                    tag="form"
          >
                    <FormGroup
                              tag="div"
                    >
                              <Input
                                        name="title"
                                        onChange={[Function]}
                                        placeholder="Title"
                                        type="text"
                                        value=""
                              />
                    </FormGroup>
                    <FormGroup
                              tag="div"
                    >
                              <Input
                                        name="body"
                                        onChange={[Function]}
                                        placeholder="Add your article"
                                        rows="20"
                                        type="textarea"
                                        value=""
                              />
                    </FormGroup>
                    
                    <div
                              className="clearfix"
                    >
                              <Button
                                        color="primary float-right"
                                        disabled={false}
                                        id="create-article-btn"
                                        size="lg"
                                        tag="button"
                              >
                                        Save
                              </Button>
                    </div>
          </Form>
</div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "resourceTitle": "Create New Article",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <Form
            autoComplete="off"
            onSubmit={[Function]}
            tag="form"
>
            <FormGroup
                        tag="div"
            >
                        <Input
                                    name="title"
                                    onChange={[Function]}
                                    placeholder="Title"
                                    type="text"
                                    value=""
                        />
            </FormGroup>
            <FormGroup
                        tag="div"
            >
                        <Input
                                    name="body"
                                    onChange={[Function]}
                                    placeholder="Add your article"
                                    rows="20"
                                    type="textarea"
                                    value=""
                        />
            </FormGroup>
            
            <div
                        className="clearfix"
            >
                        <Button
                                    color="primary float-right"
                                    disabled={false}
                                    id="create-article-btn"
                                    size="lg"
                                    tag="button"
                        >
                                    Save
                        </Button>
            </div>
</Form>,
          "className": "container",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "autoComplete": "off",
            "children": Array [
              <FormGroup
                tag="div"
>
                <Input
                                name="title"
                                onChange={[Function]}
                                placeholder="Title"
                                type="text"
                                value=""
                />
</FormGroup>,
              <FormGroup
                tag="div"
>
                <Input
                                name="body"
                                onChange={[Function]}
                                placeholder="Add your article"
                                rows="20"
                                type="textarea"
                                value=""
                />
</FormGroup>,
              "",
              <div
                className="clearfix"
>
                <Button
                                color="primary float-right"
                                disabled={false}
                                id="create-article-btn"
                                size="lg"
                                tag="button"
                >
                                Save
                </Button>
</div>,
            ],
            "onSubmit": [Function],
            "tag": "form",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": <Input
                  name="title"
                  onChange={[Function]}
                  placeholder="Title"
                  type="text"
                  value=""
/>,
                "tag": "div",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "name": "title",
                  "onChange": [Function],
                  "placeholder": "Title",
                  "type": "text",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": <Input
                  name="body"
                  onChange={[Function]}
                  placeholder="Add your article"
                  rows="20"
                  type="textarea"
                  value=""
/>,
                "tag": "div",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "name": "body",
                  "onChange": [Function],
                  "placeholder": "Add your article",
                  "rows": "20",
                  "type": "textarea",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            "",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <Button
                  color="primary float-right"
                  disabled={false}
                  id="create-article-btn"
                  size="lg"
                  tag="button"
>
                  Save
</Button>,
                "className": "clearfix",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "Save",
                  "color": "primary float-right",
                  "disabled": false,
                  "id": "create-article-btn",
                  "size": "lg",
                  "tag": "button",
                },
                "ref": null,
                "rendered": "Save",
                "type": [Function],
              },
              "type": "div",
            },
          ],
          "type": [Function],
        },
        "type": "div",
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Header
            resourceTitle="Create New Article"
/>,
          <div
            className="container"
>
            <Form
                        autoComplete="off"
                        onSubmit={[Function]}
                        tag="form"
            >
                        <FormGroup
                                    tag="div"
                        >
                                    <Input
                                                name="title"
                                                onChange={[Function]}
                                                placeholder="Title"
                                                type="text"
                                                value=""
                                    />
                        </FormGroup>
                        <FormGroup
                                    tag="div"
                        >
                                    <Input
                                                name="body"
                                                onChange={[Function]}
                                                placeholder="Add your article"
                                                rows="20"
                                                type="textarea"
                                                value=""
                                    />
                        </FormGroup>
                        
                        <div
                                    className="clearfix"
                        >
                                    <Button
                                                color="primary float-right"
                                                disabled={false}
                                                id="create-article-btn"
                                                size="lg"
                                                tag="button"
                                    >
                                                Save
                                    </Button>
                        </div>
            </Form>
</div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "resourceTitle": "Create New Article",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <Form
              autoComplete="off"
              onSubmit={[Function]}
              tag="form"
>
              <FormGroup
                            tag="div"
              >
                            <Input
                                          name="title"
                                          onChange={[Function]}
                                          placeholder="Title"
                                          type="text"
                                          value=""
                            />
              </FormGroup>
              <FormGroup
                            tag="div"
              >
                            <Input
                                          name="body"
                                          onChange={[Function]}
                                          placeholder="Add your article"
                                          rows="20"
                                          type="textarea"
                                          value=""
                            />
              </FormGroup>
              
              <div
                            className="clearfix"
              >
                            <Button
                                          color="primary float-right"
                                          disabled={false}
                                          id="create-article-btn"
                                          size="lg"
                                          tag="button"
                            >
                                          Save
                            </Button>
              </div>
</Form>,
            "className": "container",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "autoComplete": "off",
              "children": Array [
                <FormGroup
                  tag="div"
>
                  <Input
                                    name="title"
                                    onChange={[Function]}
                                    placeholder="Title"
                                    type="text"
                                    value=""
                  />
</FormGroup>,
                <FormGroup
                  tag="div"
>
                  <Input
                                    name="body"
                                    onChange={[Function]}
                                    placeholder="Add your article"
                                    rows="20"
                                    type="textarea"
                                    value=""
                  />
</FormGroup>,
                "",
                <div
                  className="clearfix"
>
                  <Button
                                    color="primary float-right"
                                    disabled={false}
                                    id="create-article-btn"
                                    size="lg"
                                    tag="button"
                  >
                                    Save
                  </Button>
</div>,
              ],
              "onSubmit": [Function],
              "tag": "form",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": <Input
                    name="title"
                    onChange={[Function]}
                    placeholder="Title"
                    type="text"
                    value=""
/>,
                  "tag": "div",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "name": "title",
                    "onChange": [Function],
                    "placeholder": "Title",
                    "type": "text",
                    "value": "",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": <Input
                    name="body"
                    onChange={[Function]}
                    placeholder="Add your article"
                    rows="20"
                    type="textarea"
                    value=""
/>,
                  "tag": "div",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "name": "body",
                    "onChange": [Function],
                    "placeholder": "Add your article",
                    "rows": "20",
                    "type": "textarea",
                    "value": "",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": [Function],
              },
              "",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <Button
                    color="primary float-right"
                    disabled={false}
                    id="create-article-btn"
                    size="lg"
                    tag="button"
>
                    Save
</Button>,
                  "className": "clearfix",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "children": "Save",
                    "color": "primary float-right",
                    "disabled": false,
                    "id": "create-article-btn",
                    "size": "lg",
                    "tag": "button",
                  },
                  "ref": null,
                  "rendered": "Save",
                  "type": [Function],
                },
                "type": "div",
              },
            ],
            "type": [Function],
          },
          "type": "div",
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
